metadata-ingestion/src/datahub/ingestion/source/qlik_sense/qlik_sense.py

```diff
@@ -290,75 +301,83 @@ def _gen_sheets_workunit(
-    ) -> MetadataWorkUnit:
-        dataset_identifier = self._get_app_dataset_identifier(dataset)
-        dataset_urn = self._gen_qlik_dataset_urn(dataset_identifier)
-        upstream_dataset_platform_detail = (
-            self.config.data_connection_to_platform_instance.get(
-                dataset.dataconnectorName,
-                PlatformDetail(
-                    env=self.config.env, platform_instance=self.config.platform_instance
+    def _gen_app_table_upstream_lineage(
+        self, dataset_urn: str, table: QlikTable
+    ) -> Optional[MetadataWorkUnit]:
+        upstream_dataset_urn: Optional[str] = None
+        if table.type == BoxType.BLACKBOX:
+            upstream_dataset_platform_detail = (
+                self.config.data_connection_to_platform_instance.get(
+                    table.dataconnectorName,
+                    PlatformDetail(
+                        env=self.config.env,
+                        platform_instance=self.config.platform_instance,
+                    ),
+                )
+            )
+            upstream_dataset_urn = builder.make_dataset_urn_with_platform_instance(
+                name=f"{table.databaseName}.{table.schemaName}.{table.tableName}",
+                platform=KNOWN_DATA_PLATFORM_MAPPING.get(
+                    table.dataconnectorPlatform, table.dataconnectorPlatform
                 ),
+                env=upstream_dataset_platform_detail.env,
+                platform_instance=upstream_dataset_platform_detail.platform_instance,
             )
-        )
-
-        upstream_dataset_urn = builder.make_dataset_urn_with_platform_instance(
-            name=dataset_identifier,
-            platform=KNOWN_DATA_PLATFORM_MAPPING.get(
-                dataset.dataconnectorPlatform, dataset.dataconnectorPlatform
-            ),
-            env=upstream_dataset_platform_detail.env,
-            platform_instance=upstream_dataset_platform_detail.platform_instance,
-        )
-
-        return MetadataChangeProposalWrapper(
-            entityUrn=dataset_urn,
-            aspect=UpstreamLineage(
-                upstreams=[
-                    Upstream(dataset=upstream_dataset_urn, type=DatasetLineageType.COPY)
-                ]
-            ),
-        ).as_workunit()
-
-    def _gen_app_dataset_properties(self, dataset: QlikAppDataset) -> MetadataWorkUnit:
-        dataset_urn = self._gen_qlik_dataset_urn(
-            self._get_app_dataset_identifier(dataset)
-        )
+        elif table.type == BoxType.LOADFILE:
+            upstream_dataset_urn = self._gen_qlik_dataset_urn(
+                f"{table.spaceId}.{table.databaseName}"
+            )
+        if upstream_dataset_urn:
+            return MetadataChangeProposalWrapper(
+                entityUrn=dataset_urn,
+                aspect=UpstreamLineage(
+                    upstreams=[
+                        Upstream(
+                            dataset=upstream_dataset_urn, type=DatasetLineageType.COPY
```

Comment:
can we also generate fineGrainedLineage?

Reply:
We can consider as one to one mapping of tables and columns and can generare fineGrainedLineage.